src/executables/srpd_aging.c

```diff
@@ -0,0 +1,403 @@
+                            goto cleanup;
+                        }
+
+                        /* Remove a file from notif folder */
+                        if (remove(remove_str)) {
+                            SRPLG_LOG_ERR("srpd_aging", "Removing a file failed.");
+                        }
+                    }
+
+                    /* Build moving args */
+                } else {
+                    if (asprintf(&arg1, "%s%s", notif_dir_name, dir->d_name) == -1) {
+                        goto cleanup;
+                    }
+                    if (asprintf(&arg2, "%s%s", opts->archive, dir->d_name) == -1) {
+                        goto cleanup;
+                    }
+
+                    /* Move a file to archive folder */
+                    if ((rc = rename(arg1, arg2)) == -1) {
+                        SRPLG_LOG_ERR("srpd_aging", "Moving a file failed.");
+                    } else {
+                        opts->archived_notif_count++;
+                    }
+                }
+
+                /* Reset commands for next folders */
+                rc = 0;
+                free(arg1);
+                free(arg2);
+                free(remove_str);
+                arg1 = NULL;
+                arg2 = NULL;
+                remove_str = NULL;
+            }
+        }
+        closedir(d);
+        d = NULL;
+    }
+
+cleanup:
+    free(arg1);
+    free(arg2);
+    free(remove_str);
+    free(notif_dir_name);
+    free(opts->archive);
+    closedir(d);
+    return NULL;
+}
+
+/**
+ * @brief Attempts to join thread.
+ *
+ * @return 0 on success.
+ * @return -1 on failure.
+ */
+static int
+srpd_join_thread(srpd_aging_opts_t *opts)
+{
+    int rc = 0;
+
+    opts->running = 0;
+    if ((rc = pthread_join(opts->tid, NULL))) {
+        SRPLG_LOG_ERR("srpd_aging", "pthread_join failed (%s).", sr_strerror(rc));
+    }
+
+    return rc;
+}
+
+int
+srpd_notif_cb(sr_session_ctx_t *session, uint32_t sub_id, const char *module_name, const char *xpath,
+        sr_event_t event, uint32_t request_id, void *private_data)
+{
+    int rc = 0, t_creat = 0;
+    char *time_unit = NULL;
+    time_t time_value;
+    sr_change_iter_t *iter;
+    sr_change_oper_t oper;
+    const struct lyd_node *node;
+    srpd_aging_opts_t *opts = (srpd_aging_opts_t *)private_data;
+
+    (void) sub_id;
+    (void) module_name;
+    (void) xpath;
+    (void) event;
+    (void) request_id;
+
+    if (opts->finish) {
+        return srpd_join_thread(opts);
+    }
+
+    sr_get_changes_iter(session, "/sysrepo-plugind:sysrepo-plugind/notif-datastore/rotation/enabled//.", &iter);
+
+    while ((rc = sr_get_change_tree_next(session, iter, &oper, &node, NULL, NULL, NULL)) == SR_ERR_OK) {
+
+        if (!strcmp(node->schema->name, "enabled")) {
+            if (oper == SR_OP_CREATED) {
+                t_creat = 1;
+            } else if (oper == SR_OP_DELETED) {
+                rc = srpd_join_thread(opts);
```

Comment:
If continuing after an error, at least add a comment explaining that it is intentional.

Reply:
Done